### ISSUE TITLE
PowerConnect 54xx and 55xx

### ISSUE THREAD

**Issue by MrRJ45** [ytti/oxidized#89]
We have a large amount of these switches, and it looks like model/powerconnect.rb used to use:

post_login "terminal datadump"
pre_logout "exit"

This was changed some time ago to this, which is suitable for some models of PowerConnect but not the 54xx and 55xx.

post_login "terminal length 0"
pre_logout "logout"

I have tried to extract the model number from "show system" so the module can decide which commands to use, but I wasnt able to get this to work correctly.

**Comment by MrRJ45** [comment]
I should mention I have added the lines back to powerconnect.rb and that works fine, but its not an ideal solution depending on the model of switches people are using :)

**Comment by ytti** [comment]
@emjemj @nopedial 

So there isn't single command which works for all models we're using? It should be indeed possible to do this conditionally, but of course it would be best, if we can find commands which are platform agnostic.
Potentially we can also avoid conditions, and just send both, allowing one of them to fail?

**Comment by emjemj** [comment]
At least on my powerconnects, terminal datadump does not work.

Sending both should work, would the potential error message get saved along with the configuration or just be silently dropped?

**Comment by nopedial** [comment]
I believe it would get saved, as it'd be seen just as regular output. I personally do not have Powerconnects anymore to test..

**Comment by ytti** [comment]
@emjemj IIRC i think those are non-capturing commands. Only 'cmd X' captures, and even those can be made non-capturing.

**Comment by emjemj** [comment]
I'll give it a try.

**Comment by MrRJ45** [comment]
I have both versions of the commands in powerconnect.rb and it doesnt show them in the config, that is for a 5548.

I have noticed that "show system" isnt recorded in the configuration, to fix this I commented out line 23:

cfg = cfg[0..-28]<<" "

**Comment by MrRJ45** [comment]
I have noticed that quite a few of the Dell switches show the uptime in the following format:

Unit     Up time
---- ---------------
 1     98,10:29:08

This isn't removed by powerconnect.rb as it spans multiple lines. To fix, this, I've changed to the following and this works well.

  cmd 'show system' do |cfg|
    # We should find the model of PowerConnect switch this is
    @model = $1 if cfg.match /Power[C|c]onnect (\d{4}[A-Z]?)/
    clean cfg
  end

  def clean cfg
    out = []
    skip_lines = 0
    cfg.each_line do |line|
       if line.match /Up\sTime|Temperature|Power Supplies/i
           # For 34xx, 54xx, and 55xx models we should skip the next 3 lines
           if @model =~ /^34(24|48)$|^54(24|48)$|^55(24|48)$/
              skip_lines = 3
           # 8024F has verbose output for Temperature and Power Supplies
           elsif @model == '8024F'
              skip_lines = 9
           else
              next
           end
       end
       # If we have lines to skip, we should and decrement the counter
       if skip_lines > 0
          skip_lines -= 1
          next
       end
       out << line.strip
    end
    out = comment out.join "\n"
    out << "\n"
  end

Is there a better way to do this in Ruby? :)

**Comment by ytti** [comment]
Plenty of ways, of course. If you're doing anything more complex 'StringScanner' is good waya to work with strings.
If it is always in same line position, something like string = (string.lines[0..4] + string.lines[6..-1]).to_s might do the trick, or we can discover the offset:

```
ytti@ytti ~> pry
[1] pry(main)> str = "foo\nuptime\nmoar\nstill moar\nquux\n"                               
=> "foo\nuptime\nmoar\nstill moar\nquux\n"
[2] pry(main)> lines = str.lines                                                           
=> ["foo\n", "uptime\n", "moar\n", "still moar\n", "quux\n"]
[3] pry(main)> uptime_start = lines.find_index { |line| line.match /^uptime/ }             
=> 1
[4] pry(main)> (lines[0..uptime_start-1] + lines[uptime_start+3..-1]).join                 
=> "foo\nquux\n"
[5] pry(main)>                                                                             
```

**Comment by MrRJ45** [comment]
I will have a play with my code, as ignoring 3 lines is correct for a 5548 which is a single switch. For each switch in a stack there is an additional line, so it should work in that scenario.

I think best idea is to keep a similar idea, but set something so we ignore the lines, until we hit a blank line so we know that block has ended.

**Comment by ytti** [comment]
StringScanner might be solution for anything non-trivial.

Optionally, you could opt to
a) only select lines which you want string.lines.select { |line| some_test_for_interest }.join
b) delete lines which are of no interest string.lines.delete_if { |line| some_test_for_uninterest}.join

And for clarity, longer ones in multimple line
```
lines.delete_if do |line|
   something
end.join
```

**Comment by MrRJ45** [comment]
I've made this a bit nicer, looking for the next blank line. That way whether the switch is part of a stack or not this will work fine. I've tested this with all the PowerConnect's we have and it works with no problems.

I'll submit a pull request for this in a minute :)

**Comment by MrRJ45** [comment]
Closed by #96

**Comment by ytti** [comment]
closed by #96

**Comment by ytti** [comment]
closed by #96
